solutions/usaco-1159.mdx

```diff
@@ -0,0 +1,142 @@
+				adj.get(a).add(b);
+				adj.get(b).add(a);
+			}
+
+			// stores the component each node is in
+			int[] visited = new int[n];
+			// stores a list of nodes for each component
+			List<List<Integer>> components = new ArrayList<>();
+
+			int numComponents = 0;
+			Arrays.fill(visited, -1);
+			for (int i = 0; i < n; i++) {
+				// run DFS on node if it hasn't been visited
+				if (visited[i] == -1) {
+					components.add(new ArrayList<>());  // add new component
+					Stack<Integer> stack = new Stack<>();
+					stack.push(i);
+					while (!stack.isEmpty()) {
+						int curr = stack.pop();
+						if (visited[curr] != -1) continue;
+						// add node to component
+						components.get(numComponents).add(curr);
+						// mark node as visited
+						visited[curr] = numComponents;
+
+						for (int neighbor : adj.get(curr)) {
+							stack.push(neighbor);
+						}
+					}
+					numComponents++;
+				}
+			}
+			// sort nodes in each component by position
+			for (List<Integer> component : components) {
+				Collections.sort(component);
+			}
+
+			// the components containing the two barns
+			List<Integer> barn1 = components.get(visited[0]);
+			List<Integer> barn2 = components.get(visited[n - 1]);
+			// min dist between each intermediate component and the two barns
+			long[] dist1 = new long[numComponents];
+			long[] dist2 = new long[numComponents];
+			Arrays.fill(dist1, Integer.MAX_VALUE);
+			Arrays.fill(dist2, Integer.MAX_VALUE);
+
+			// use two pointer to populate dist1
+			int barn1Index = 0;
+			for (int i = 0; i < n; i++) {
+				int dist = Math.abs(barn1.get(barn1Index) - i);
+
+				/*
+				 * if the distance between i and the current field in barn1's
+				 * component is larger than the distance between i and the next
+				 * field in barn1's component, increment barn1Index until this
+				 * no longer holds true
+				 */
+				while (
+					// make sure we don't go out of bounds
+					barn1Index < barn1.size() - 1
+					&& Math.abs(barn1.get(barn1Index + 1) - i) < dist
+				) {
+					barn1Index++;
+				}
+
+				/*
+				 * we've now found the field in barn 1's component that's closest to i
+				 *
+				 * we can use this to update the minimum distance between field i's
+				 * component and barn 1's component
+				 */
+				dist1[visited[i]] = Math.min(dist, dist1[visited[i]]);
+			}
+
+			// use two pointer to populate dist2
+			int barn2Index = 0;
+			for (int i = 0; i < n; i++) {
+				int dist = Math.abs(barn2.get(barn2Index) - i);
+
+				while (
+					barn2Index < barn2.size() - 1
+					&& Math.abs(barn2.get(barn2Index + 1) - i) < dist
+				) {
+					barn2Index++;
+				}
+				dist2[visited[i]] = Math.min(dist, dist2[visited[i]]);
+			}
+
+			// calculate minimum cost
+			long min = Long.MAX_VALUE;
+			for (int i = 0; i < numComponents; i++) {
+				long cost = dist1[i] * dist1[i] + dist2[i] * dist2[i];
+				min = Math.min(min, cost);
+			}
+			io.println(min);
+		}
+		io.close();
+	}
+
+	// CodeSnip{Kattio}
```

Comment:
not related to this PR specifically but is there a reason having `Kattio` as a `static class` inside of `ConnectingTwoBarns` is preferred to having `Kattio` outside of `ConnectingTwoBarns`?